orttraining/orttraining/python/training/torchdynamo/ort_backend.py

```diff
@@ -0,0 +1,519 @@
+        self.input_names: Dict[torch.fx.GraphModule, Tuple[str]] = {}
+        # Similar to self.input_names, but for outputs of the graph.
+        self.output_names: Dict[torch.fx.GraphModule, Tuple[str]] = {}
+        # self.input_devices[mod] contains devices of inputs fed to mod.forward (excluding self).
+        # self.input_devices[mod][i] is the i-th positional input's device.
+        self.input_devices: Dict[torch.fx.GraphModule, Tuple[ORTC.OrtDevice]] = {}
+        # Similar to self.input_devices, but for outputs of the graph.
+        self.output_devices: Dict[torch.fx.GraphModule, Tuple[ORTC.OrtDevice]] = {}
+        # This is a debug flag. When True, this backend will compare its
+        self.assert_allclose_to_baseline: bool = False
+        # We need example outputs to determine output schema of ORT run.
+        # self.example_outputs[mod] is the outputs of mod.forward(*self.example_inputs[mod]).
+        self.example_outputs: Dict[torch.fx.GraphModule, Union[Tuple[torch.Tensor], torch.Tensor]] = {}
+
+
+class OrtBackend:
+    """A backend compiles (sub-)graphs in torch.fx.GraphModule to onnxruntime.InferenceSession calls.
+
+    The compiler entry point is OrtBackend.compile, which
+        1. partitions the original graph into supported sub-graphs (type: torch.fx.GrpahModule) and unsupported
+           sub-graphs.
+        2. For each supported sub-graph, it replaces its _wrapped_call function with _ort_accelerated_call.
+        3. Inside _ort_accelerated_call, it creates onnxruntime.InferenceSession and calls it to execute the sub-graph.
+    """
+
+    def __init__(self):
+        self.supported_ops = OrtOperatorSupport()
+        # TODO: this is a naive implementation of cache without proper guard
+        self.partitioner_cache: Dict[torch.fx.GraphModule, torch.fx.GraphModule] = {}
+        # TODO: this is a naive implementation of cache without proper guard, this will only work for identical inputs
+        self.ort_execution_info = OrtExecutionInfo()
+
+    def _ort_acclerated_call(self, graph_module: torch.fx.GraphModule, *args, **kwargs):
+        if graph_module in self.ort_execution_info.sessions:
+            # We have seen this graph before, so we can use cached objects including session.
+            onnx_session = self.ort_execution_info.sessions[graph_module]
+            input_names = self.ort_execution_info.input_names[graph_module]
+            output_names = self.ort_execution_info.output_names[graph_module]
+            input_devices = self.ort_execution_info.input_devices[graph_module]
+            output_devices = self.ort_execution_info.output_devices[graph_module]
+            prim_outputs = self.ort_execution_info.example_outputs[graph_module]
+        else:
+            # It's first time seeing such as graph. Let's make a new session
+            # (type: onnxruntime.InferenceSession) for it.
+
+            # TODO(wechi): this is a workaround for #84311 in PyTorch.
```

Comment:
```suggestion
            # TODO(wechi): this is a workaround for pytorch/pytorch#84311.
```

so that one can click on github and get to the right place